### PR TITLE
Modified returns for attributes to handle nulls

### PR DESCRIPTION
**Pull request by holmesie**
- Modified return values for attributes in each datatype in order to properly handle null cases
- Moved components attribute logic to helpers for duplicate code reduction
- Moved diagnosis attribute logic to helpers for duplicate code reduction
- Removed unnecessary getCode method from patient_characteristic.js.coffee and patient_characteristic_payer.js.coffee

### CODE REVIEW

**Comment by jbradl11** [comment]
Is there a Jira test for this?

**Comment by edeyoung** [comment]
Here's one Matt put together that is directly related to how this was found: 
https://jira.mitre.org/browse/BONNIE-1057

**Comment by edeyoung** [comment]
Seems like some unit tests putting nulls into things and passing them in to see how they behave *might* (strong might because everything is harder than it seems) be easy to gen up for this.

File: app/assets/javascripts/datatypes/assessment.js.coffee [hunk 1]
```diff
@@ -60,6 +69,8 @@ class CQL_QDM.AssessmentPerformed extends CQL_QDM.QDMDatatype
         CQL_QDM.Helpers.convertDateTime(@_result.scalar)
       else
         CQL_QDM.Helpers.formatResult(@_result)
+    else
```
Comment (losborne) [inline app/assets/javascripts/datatypes/assessment.js.coffee]: Should the above if be `if @_result?` ?

File: app/assets/javascripts/datatypes/medication.js.coffee [hunk 2]
```diff
@@ -175,13 +211,19 @@ class CQL_QDM.MedicationDischarge extends CQL_QDM.QDMDatatype
   @returns {Code}
   ###
   route: ->
-    new cql.Code(@_route?.code, @_route?.code_system)
+    if @_route
```
Comment (losborne) [inline app/assets/javascripts/datatypes/medication.js.coffee]: Should this be `@_route?`?

File: app/assets/javascripts/datatypes/medication.js.coffee [hunk 3]
```diff
@@ -224,21 +266,30 @@ class CQL_QDM.MedicationDispensed extends CQL_QDM.QDMDatatype
...
+      new cql.Quantity({unit: @_dosage['unit'], value: @_dosage['value']})
+    else
+      null
 
   ###
   @returns {Code}
   ###
   negationRationale: ->
-    new cql.Code(@_negationRationale?.code, @_negationRationale?.code_system)
+    if @_negationRationale
```
Comment (losborne) [inline app/assets/javascripts/datatypes/medication.js.coffee]: Should this be `@_negationRationale?`?

File: app/assets/javascripts/datatypes/medication.js.coffee [hunk 4]
```diff
@@ -303,25 +360,37 @@ class CQL_QDM.MedicationOrder extends CQL_QDM.QDMDatatype
...
+      new cql.Code(@_method.code, @_method.code_system)
+    else
+      null
 
   ###
   @returns {Code}
   ###
   negationRationale: ->
-    new cql.Code(@_negationRationale?.code, @_negationRationale?.code_system)
+    if @_negationRationale
```
Comment (losborne) [inline app/assets/javascripts/datatypes/medication.js.coffee]: should this be `@_negationRationale?`?

Comment (holmesie) [reply app/assets/javascripts/datatypes/medication.js.coffee]: Yes, good catch! x3

File: app/assets/javascripts/datatypes/medication.js.coffee (hunk 3 above)
Comment (holmesie) [reply app/assets/javascripts/datatypes/medication.js.coffee]: Yes, good catch! x2

File: app/assets/javascripts/datatypes/medication.js.coffee (hunk 2 above)
Comment (holmesie) [reply app/assets/javascripts/datatypes/medication.js.coffee]: Yes, good catch! x1

File: app/assets/javascripts/datatypes/assessment.js.coffee (hunk 1 above)
Comment (holmesie) [reply app/assets/javascripts/datatypes/assessment.js.coffee]: The helper takes care of @_result being null in this case, but interestingly the helper forgot a `?`, so I've added that.

File: app/assets/javascripts/datatypes/encounter.js.coffee [hunk 5]
```diff
@@ -119,27 +126,41 @@ class CQL_QDM.EncounterPerformed extends CQL_QDM.QDMDatatype
   @returns {Quantity}
   ###
   lengthOfStay: ->
-    new cql.Quantity({unit: 'days', value: @_relevantPeriodLow.differenceBetween(@_relevantPeriodHigh, 'day')?.high})
+    low = @_relevantPeriodLow
+    high = @_relevantPeriodHigh
+    if low? || high?
```
Comment (hossenlopp) [inline app/assets/javascripts/datatypes/encounter.js.coffee]: Shouldn't this only attempt to create the quantity if it has both a low and high?

File: app/assets/javascripts/datatypes/encounter.js.coffee [hunk 6]
```diff
@@ -119,27 +126,41 @@ class CQL_QDM.EncounterPerformed extends CQL_QDM.QDMDatatype
   @returns {Quantity}
   ###
   lengthOfStay: ->
-    new cql.Quantity({unit: 'days', value: @_relevantPeriodLow.differenceBetween(@_relevantPeriodHigh, 'day')?.high})
+    low = @_relevantPeriodLow
+    high = @_relevantPeriodHigh
+    if low?
```
Comment (hossenlopp) [inline app/assets/javascripts/datatypes/encounter.js.coffee]: Shouldn't this require both low and high to create the Quantity.

File: app/assets/javascripts/datatypes/immunization.js.coffee [hunk 7]
```diff
@@ -95,28 +110,43 @@ class CQL_QDM.ImmunizationOrder extends CQL_QDM.QDMDatatype
...
+    else
+      null
 
   ###
   @returns {Quantity}
   ###
   supply: ->
-    new cql.Quantity({unit: @_supply['unit'], value: @_supply['value']})
+    if @_supply?
+      new cql.Quantity({unit: @_supply['unit'], value: @_supply['value']})
```
Comment (hossenlopp) [inline app/assets/javascripts/datatypes/immunization.js.coffee]: Why aren't you doing `@_supply.unit` and `@_supply.value` here?

File: app/assets/javascripts/datatypes/medication.js.coffee [hunk 8]
```diff
@@ -33,27 +33,39 @@ class CQL_QDM.MedicationActive extends CQL_QDM.QDMDatatype
   @returns {Quantity}
   ###
   dosage: ->
-    new cql.Quantity({unit: @_dosage['unit'], value: @_dosage['value']})
+    if @_dosage?
+      new cql.Quantity({unit: @_dosage['unit'], value: @_dosage['value']})
```
Comment (hossenlopp) [inline app/assets/javascripts/datatypes/medication.js.coffee]: Why are the bracket operators being used here?

File: app/assets/javascripts/datatypes/medication.js.coffee [hunk 9]
```diff
@@ -33,27 +33,39 @@ class CQL_QDM.MedicationActive extends CQL_QDM.QDMDatatype
...
+    else
+      null
 
   ###
   @returns {Quantity}
   ###
   supply: ->
-    new cql.Quantity({unit: @_supply['unit'], value: @_supply['value']})
+    if @_supply?
+      new cql.Quantity({unit: @_supply['unit'], value: @_supply['value']})
```
Comment (hossenlopp) [inline app/assets/javascripts/datatypes/medication.js.coffee]: Bracket operators again. I know it's not new code but I'm still curious.

File: app/assets/javascripts/datatypes/medication.js.coffee [hunk 10]
```diff
@@ -92,39 +104,57 @@ class CQL_QDM.MedicationAdministered extends CQL_QDM.QDMDatatype
...
+      new cql.Quantity({unit: @_dosage['unit'], value: @_dosage['value']})
+    else
+      null
 
   ###
   @returns {Code}
   ###
   negationRationale: ->
-    new cql.Code(@_negationRationale?.code, @_negationRationale?.code_system)
+    if @negationRationale?
```
Comment (hossenlopp) [inline app/assets/javascripts/datatypes/medication.js.coffee]: This should be `@_negationRationale`.

File: app/assets/javascripts/datatypes/encounter.js.coffee (hunk 6 above)
Comment (holmesie) [reply app/assets/javascripts/datatypes/encounter.js.coffee]: `@_relevantPeriodHigh` is never null, not worth checking.

**Comment by edeyoung** [comment]
after applying this, only expected regressions remain (diagnoses and facilityLocations related regressions)

**Pull request merged**